tensorflow/core/kernels/unique_op_gpu.cu.cc

```diff
@@ -0,0 +1,468 @@
+==============================================================================*/
+
+#if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
+
+#define EIGEN_USE_GPU
+
+#include "tensorflow/core/common_runtime/gpu/gpu_event_mgr.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/kernels/gpu_prim.h"
+#include "tensorflow/core/kernels/gpu_prim_helpers.h"
+#include "tensorflow/core/lib/core/bits.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/util/gpu_kernel_helper.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+
+#if GOOGLE_CUDA
+#include "tensorflow/core/util/cuda_solvers.h"  // For ScratchSpace
+#elif TENSORFLOW_USE_ROCM
+#include "tensorflow/core/util/rocm_solvers.h"
+#endif
+
+namespace tensorflow {
+
+typedef Eigen::GpuDevice GPUDevice;
+
+namespace {
+
+// Returns true iff index is at the end of a segment (which is equivalent to the
+// beginning of the next segment).
+template <typename T, typename TIndex>
+struct SegmentIndicatorFunctor {
+  const T* __restrict__ sorted_input_ptr_;
+  SegmentIndicatorFunctor(const T* sorted_input_ptr)
+      : sorted_input_ptr_(sorted_input_ptr) {}
+  __device__ bool operator()(const TIndex& i) const {
+    return i > 0 && sorted_input_ptr_[i] != sorted_input_ptr_[i - 1];
+  }
+};
+
+template <typename TIndex>
+__global__ void ExtractFirstOccurrenceIndicesKernel(
+    int64 input_size, int64 uniq_size,
+    const TIndex* __restrict__ sorted_input_inds,
+    const TIndex* __restrict__ sorted_input_unique_ids,
+    TIndex* __restrict__ unique_input_inds, TIndex* __restrict__ segment_ends) {
+  GPU_1D_KERNEL_LOOP(i, input_size) {
+    TIndex sorted_input_unique_id = sorted_input_unique_ids[i];
+    if (i == 0 || sorted_input_unique_id != sorted_input_unique_ids[i - 1]) {
+      unique_input_inds[sorted_input_unique_id] = sorted_input_inds[i];
+      if (segment_ends) {
+        if (i == 0) {
+          // First thread writes the last element.
+          segment_ends[uniq_size - 1] = input_size;
+        } else {
+          segment_ends[sorted_input_unique_id - 1] = i;
+        }
+      }
+    }
+  }
+}
+
+// Scatters the index of the first occurrence of each unique input value to
+// unique_input_inds.
+// If segment_ends is not nullptr, it is filled with the end index of each
+// unique value's range in the sorted input (the last element is always set
+// to input_size).
+template <typename TIndex>
+Status ExtractFirstOccurrenceIndices(const GPUDevice& d, int64 input_size,
+                                     int64 uniq_size,
+                                     const TIndex* sorted_input_inds,
+                                     const TIndex* sorted_input_unique_ids,
+                                     TIndex* unique_input_inds,
+                                     TIndex* segment_ends) {
+  if (input_size == 0) return Status::OK();
```

Comment:
This can be a `CHECK` right now that you have an early exit for empty input?  Same elsewhere.